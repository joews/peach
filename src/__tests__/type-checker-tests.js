/* eslint-env jest */
const parse = require('../parser')
const typeCheck = require('../type-checker')
const { fixture } = require('./helpers')
const { clone } = require('../util')

const {
  TypeVariable,
  TypeOperator,
  FunctionType,
  NumberType,
  BooleanType
} = require('../types')

// TODO unify envs
const rootEnv = require('../interpreter').getRootEnv()
const defaultEnv = () => clone(typeCheck.getTypeEnv(rootEnv))

//
// snapshot tests for the parser
//

function testTypeCheck (code, env = defaultEnv()) {
  test(code, () => {
    const parsed = parse(code)
    const analysed = typeCheck(parsed, env)

    const types = analysed.map(e => e[0].exprType.toString())
    expect(types).toMatchSnapshot()
  })
};

function testFails (code, env = defaultEnv()) {
  test(code, () => {
    const parsed = parse(code)
    expect(() => typeCheck(parsed, env)).toThrowErrorMatchingSnapshot()
  })
}

function testFixture (fixtureName, env = defaultEnv()) {
  // assert that the no-op analyser makes no changes
  test(fixtureName, () => {
    const parsed = parse(fixture(fixtureName))
    const analysed = typeCheck(parsed, env)
    expect(analysed.map(e => e[0].exprType.toString())).toMatchSnapshot()
  })
};

testFixture('fibonacci.peach')
testFixture('function.peach')
testFixture('str.peach')
testFixture('tail-recursion.peach')
testFixture('array-destructure.peach')

// literals
testTypeCheck(`1`)
testTypeCheck(`true`)
testTypeCheck('`the`')

// def
testTypeCheck(`
  x = 1
  y = true
  x
  y
`)

testTypeCheck(`
  x = \`arf\`
  x
`)

// lambda
testTypeCheck(`(a => 1)`)

// array
testTypeCheck(`[1 2 3]`)
testFails(`[1 2 false]`)

// if
testTypeCheck(`if (true) 1 else 2`)

// branches evaluate to different types
testFails(`(if (true) 1 else \`two\`)`)

// non-boolean condition
testFails(`if (1) 1 else 2`)

// test with some pre-defined types and symbols
// run adapted version of Rob Smallshire's python implementation for a sanity test:
// http://smallshire.org.uk/sufficientlysmall/2010/04/11/a-hindley-milner-type-inference-implementation-in-python/
// (adapted because Peach doesn't have `let` yet; use `def` and test in the enclosing environment)

// the peach type checker works over nodes with an `exprType` property.
// helper for creating nodes for synthetic type tests
function typed (type) {
  return {
    exprType: type
  }
}

// simulate a user-defined type
class PairType extends TypeOperator {
  constructor (a, b) {
    super('*', [a, b])
  }

  static of (name, typeArgs) {
    return new PairType(...typeArgs)
  }

  toString () {
    return `[${this.typeArgs.join(this.name)}]`
  }
}

const A = typed(new TypeVariable())
const B = typed(new TypeVariable())
const AB = typed(new PairType(A, B))

const testEnv = () => ({
  // No unit type yet, so all functions take exactly one argument
  // Number -> Number
  inc: typed(new FunctionType(NumberType, NumberType)),

  // // Number -> Boolean
  zero: typed(new FunctionType(NumberType, BooleanType)),

  add: typed(new FunctionType(typed(NumberType), typed(new FunctionType(typed(NumberType), typed(NumberType))))),
  sub: typed(new FunctionType(typed(NumberType), typed(new FunctionType(typed(NumberType), typed(NumberType))))),

  // Number -> Number -> Boolean
  addZero: typed(new FunctionType(typed(NumberType), typed(new FunctionType(typed(NumberType), typed(BooleanType))))),

  // A -> B -> A*B
  pair: typed(new FunctionType(A, typed(new FunctionType(B, AB))))
})

// testTypeCheck(`inc`, testEnv())
testTypeCheck(`(inc 1)`, testEnv())

testTypeCheck(`zero`, testEnv())
testTypeCheck(`(zero 0)`, testEnv())
testTypeCheck(`(addZero 1)`, testEnv())
testTypeCheck(`((addZero 1) 2)`, testEnv())

testTypeCheck(`(pair 1)`, testEnv())
testTypeCheck(`((pair 1) 2)`, testEnv())
testTypeCheck(`
  id = x => x
  (id 3)
  (id \`hello\`)
`, testEnv())

testTypeCheck(`
 fib =
    0 => 1
    1 => 1
    x => ((add (fib ((sub x) 1))) (fib ((sub x) 2)))
`, testEnv())

// type mismatch in recursive call
testFails(`
  fibc =
    0 => 1
    1 => 1
    x => ((add (fibc true)) (fibc ((sub x) 3)))
`, testEnv())

// type mismatch in patterns
testFails(`0 => 0, true => 1`, testEnv())

// type mismatch in return
testFails(`0 => 0, 1 => true`, testEnv())

// type mismatch in arguments to `x`, because function arguments are non-generic
testFails(`x => ((pair (x 3)) (x true))`, testEnv())

// polymorphic function call
testTypeCheck(`
  id = x => x
  ((pair (id 3)) (id true))
`, testEnv())

// recursive function arguments can't be unified with no further type information
testFails(`f => (f f)`)

// ...but it's ok when there is more information available
testTypeCheck(`
  g = f => 5
  (g g)
`)

// // TODO I can't run this test from Rob Smallshire's implementation
// yet because function bodies are a single expression and Peach doesn't have
// `let` yet.
// generic and non-generic variables
// testTypeCheck(`
//   (g =>
//     (def f x => g)
//     ((pair (f 3)) (f true))
//   )
// `, testEnv())

// function composition
testTypeCheck(`f => g => arg => (g (f arg))`)

// destructured aguments
testTypeCheck(`[1|list] => list`)
testTypeCheck(`([1|[2|t]] => t)`)
testTypeCheck(`([x|[y|t]] => [[x y] t])`)
testFails(`([1|[true|t]] => t)`)
testFails(`([h|[true|t]] => [[h 1] t])`)

// curried function calls
testTypeCheck(`
  list = (a, b, c) => [a b c]
  (list 1)
  (list 1 2)
  (list 1 2 3)
  ((list 1) 2 3)
  (list true false true)
`)

testFails(`
  list = (a, b, c) => [a b c]
  (list 1 true)
`)

testTypeCheck(`
  f =
    (1 2) => [9 9]
    (a b) => [a b]

  (f 1)
  (f 1 2)
  ((f 1) 2)
  (f 3)
  (f 3 4)
  ((f 3) 4)
`)

// FIXME #6 - extra parens needed to immediately invoke a no-arg function
testTypeCheck('((() => `look ma no args`))')
