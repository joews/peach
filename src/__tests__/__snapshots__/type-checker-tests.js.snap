exports[`test 
  (def f
    (1 2) => [9 9]
    (a b) => [a b])

  (f 1)
  (f 1 2)
  ((f 1) 2)
  (f 3)
  (f 3 4)
  ((f 3) 4)
 1`] = `
Array [
  "Number -> Number -> Vector<Number>",
  "Number -> Vector<Number>",
  "Vector<Number>",
  "Vector<Number>",
  "Number -> Vector<Number>",
  "Vector<Number>",
  "Vector<Number>",
]
`;

exports[`test 
  (def fib
    0 => 1
    1 => 1
    x => ((add (fib ((sub x) 1))) (fib ((sub x) 2))))
 1`] = `
Array [
  "Number -> Number",
]
`;

exports[`test 
  (def fibc
    0 => 1
    1 => 1
    x => ((add (fibc true)) (fibc ((sub x) 3))))
 1`] = `"Type error: type mismatch"`;

exports[`test 
  (def g f => 5)
  (g g)
 1`] = `
Array [
  "M -> Number",
  "Number",
]
`;

exports[`test 
  (def id x => x)
  ((pair (id 3)) (id true))
 1`] = `
Array [
  "K -> K",
  "L",
]
`;

exports[`test 
  (def list (a, b, c) => [a b c])
  (list 1 true)
 1`] = `"Type error: type mismatch"`;

exports[`test 
  (def list (a, b, c) => [a b c])
  (list 1)
  (list 1 2)
  (list 1 2 3)
  ((list 1) 2 3)
  (list true false true)
 1`] = `
Array [
  "R -> R -> R -> Vector<R>",
  "Number -> Number -> Vector<Number>",
  "Number -> Vector<Number>",
  "Vector<Number>",
  "Vector<Number>",
  "Vector<Boolean>",
]
`;

exports[`test ((() => \`look ma no args\`)) 1`] = `
Array [
  "String",
]
`;

exports[`test ((addZero 1) 2) 1`] = `
Array [
  "G",
]
`;

exports[`test ((pair 1) 2) 1`] = `
Array [
  "I",
]
`;

exports[`test (? 1 1 false 2) 1`] = `"Type error: type mismatch"`;

exports[`test (? true 1 false \`two\`) 1`] = `"Type error: type mismatch"`;

exports[`test (? true 1 false 2) 1`] = `
Array [
  "Number",
]
`;

exports[`test ([1|[2|t]] => t) 1`] = `
Array [
  "Vector<Number> -> Vector<Number>",
]
`;

exports[`test ([1|[true|t]] => t) 1`] = `"Type error: type mismatch"`;

exports[`test ([h|[true|t]] => [[h 1] t]) 1`] = `"Type error: type mismatch"`;

exports[`test ([x|[y|t]] => [[x y] t]) 1`] = `
Array [
  "Vector<Q> -> Vector<Vector<Q>>",
]
`;

exports[`test (a => 1) 1`] = `
Array [
  "E -> Number",
]
`;

exports[`test (addZero 1) 1`] = `
Array [
  "F",
]
`;

exports[`test (def id x => x) (id 3) (id \`hello\`) 1`] = `
Array [
  "J -> J",
  "Number",
  "String",
]
`;

exports[`test (def x \`arf\`) x 1`] = `
Array [
  "String",
  "String",
]
`;

exports[`test (def x 1) (def y true) x y 1`] = `
Array [
  "Number",
  "Boolean",
  "Number",
  "Boolean",
]
`;

exports[`test (inc 1) 1`] = `
Array [
  "Number",
]
`;

exports[`test (pair 1) 1`] = `
Array [
  "H",
]
`;

exports[`test (zero 0) 1`] = `
Array [
  "Boolean",
]
`;

exports[`test [1 2 3] 1`] = `
Array [
  "Vector<Number>",
]
`;

exports[`test [1 2 false] 1`] = `"Type error: type mismatch"`;

exports[`test [1|list] => list 1`] = `
Array [
  "Vector<Number> -> Vector<Number>",
]
`;

exports[`test \`the\` 1`] = `
Array [
  "String",
]
`;

exports[`test 0 => 0, 1 => true 1`] = `"Type error: type mismatch"`;

exports[`test 0 => 0, true => 1 1`] = `"Type error: type mismatch"`;

exports[`test 1 1`] = `
Array [
  "Number",
]
`;

exports[`test f => (f f) 1`] = `"Type error: recursive unification"`;

exports[`test f => g => arg => (g (f arg)) 1`] = `
Array [
  "(N -> O) -> (O -> P) -> N -> P",
]
`;

exports[`test fibonacci.peach 1`] = `
Array [
  "Number -> Number",
  "Vector<Number>",
]
`;

exports[`test function.peach 1`] = `
Array [
  "A -> A",
]
`;

exports[`test str.peach 1`] = `
Array [
  "Vector<String>",
]
`;

exports[`test tail-recursion.peach 1`] = `
Array [
  "Number -> Number -> Number",
  "Number -> Number",
  "Number",
]
`;

exports[`test true 1`] = `
Array [
  "Boolean",
]
`;

exports[`test vector-destructure.peach 1`] = `
Array [
  "Vector<B> -> B",
  "Vector<C> -> C",
  "Vector<D> -> D",
  "Vector<Number>",
  "Vector<Number>",
]
`;

exports[`test x => ((pair (x 3)) (x true)) 1`] = `"Type error: type mismatch"`;

exports[`test zero 1`] = `
Array [
  "Number -> Boolean",
]
`;
